substrate/frame/proxy/src/lib.rs

```diff
@@ -508,10 +511,46 @@ pub mod pallet {
 			})
 			.map_err(|_| Error::<T>::Unannounced)?;
 
-			Self::do_proxy(def, real, *call);
+			// The return value of the proxied call is ignored.
+			let _ = Self::do_proxy(def, real, *call);
 
 			Ok(())
 		}
+
+		/// Dispatch the given `call` from an account that the sender is authorised for through
+		/// `add_proxy`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// Any error from `call` will propagate up, otherwise this will emit a `ProxyExecuted`
+		/// event with `Ok` result.
+		///
+		/// Parameters:
+		/// - `real`: The account that the proxy will make a call on behalf of.
+		/// - `force_proxy_type`: Specify the exact proxy type to be used and checked for this call.
+		/// - `call`: The call to be made by the `real` account.
+		#[pallet::call_index(10)]
+		#[pallet::weight({
+			let di = call.get_dispatch_info();
+			(T::WeightInfo::proxy_propagate_error(T::MaxProxies::get())
+				// AccountData for inner call origin accountdata.
+				.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+				.saturating_add(di.weight),
+			di.class)
+		})]
+		pub fn proxy_propagate_error(
+			origin: OriginFor<T>,
+			real: AccountIdLookupOf<T>,
+			force_proxy_type: Option<T::ProxyType>,
+			call: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let real = T::Lookup::lookup(real)?;
+			let def = Self::find_proxy(&real, &who, force_proxy_type)?;
+			ensure!(def.delay.is_zero(), Error::<T>::Unannounced);
+
+			Self::do_proxy(def, real, *call)
```

Comment:
Actually this brings up a nice optimization. Currently the `proxy` call is not refunding any weight. So, what we would need to do is to compare the pre weight with the post weight and return the excess. 

https://github.com/paritytech/polkadot-sdk/blob/0d661eac46a02f3fd7478bb703fe1dce8091891b/substrate/frame/ranked-collective/src/lib.rs#L583-L586 similar to this one, but we also need to take the weight from the inner call into account.